### PR TITLE
Add Teleportation gun/grenade/rifle (finished)

### PR DESCRIPTION
**Pull request by Ryozuki**
As i made some changes and force-pushed i can't reopen the old pr: https://github.com/ddnet/ddnet/pull/1075

Features:
- Teleport gun: Teleports you where it collided (tees included, in a special way). Doesn't teleport if the collided tile is an untele tile.
- Teleport grenade: Teleports you where it collided (tees included, in a special way). Doesn't teleport if the collided tile is an untele tile.
- Teleport rifle (laser): Teleports you where the laser runs out of energy or if it collides with a tee. Doesn't teleport if the laser bounced in an untele tile.

### CODE REVIEW

File: src/game/server/entities/projectile.cpp [hunk 1]
```diff
@@ -168,6 +170,33 @@ void CProjectile::Tick()
 		}
 		else if(pTargetChr && m_Freeze && ((m_Layer == LAYER_SWITCH && GameServer()->Collision()->m_pSwitchers[m_Number].m_Status[pTargetChr->Team()]) || m_Layer != LAYER_SWITCH))
 			pTargetChr->Freeze();
+
+		if (pOwnerChar && ColPos && !GameLayerClipped(ColPos) &&
+			((m_Type == WEAPON_GRENADE && pOwnerChar->m_HasTeleGrenade) || (m_Type == WEAPON_GUN && pOwnerChar->m_HasTeleGun)))
+		{
+			int MapIndex = GameServer()->Collision()->GetPureMapIndex(pTargetChr ? pTargetChr->m_Pos.x : ColPos.x,
```
Comment (timakro) [inline src/game/server/entities/projectile.cpp]: `GetPureMapIndex(pTargetChr ? pTargetChr->m_Pos : ColPos)`

File: src/game/server/entities/projectile.cpp [hunk 2]
```diff
@@ -168,6 +170,33 @@ void CProjectile::Tick()
...
+				pTargetChr ? pTargetChr->m_Pos.y : ColPos.y);
+			int TileIndex = GameServer()->Collision()->GetTileIndex(MapIndex);
+			int TileFIndex = GameServer()->Collision()->GetFTileIndex(MapIndex);
+
+			if ((m_Type == WEAPON_GUN || m_Type == WEAPON_GRENADE) && (TileIndex != TILE_NO_TELE_GUN && TileFIndex != TILE_NO_TELE_GUN))
```
Comment (timakro) [inline src/game/server/entities/projectile.cpp]: You checked for weapon being gun or grenade in an outer if already.

File: src/game/server/entity.cpp [hunk 3]
```diff
@@ -54,3 +54,43 @@ bool CEntity::GameLayerClipped(vec2 CheckPos)
...
+	if (!GameServer()->Collision()->TestBox(*pOutPos, vec2(28.0f, 28.0f)))
+		return true;
+
+	return false;
+}
+
+bool CEntity::GetNearestAirPosPlayer(vec2 PlayerPos, vec2* OutPos)
+{
+	int dist = 5;
+	for (; dist >= -1; dist--)
```
Comment (def-) [inline src/game/server/entity.cpp]: why not `for (int dist = 5; dist >= -1; dist--)`?

**Comment by def-** [comment]
I'm fine with this. Any opinions @Learath2 @heinrich5991 ?

File: src/game/server/entities/projectile.cpp [hunk 4]
```diff
@@ -121,7 +120,10 @@ void CProjectile::Tick()
 	if(m_Owner >= 0)
 		pOwnerChar = GameServer()->GetPlayerChar(m_Owner);
 
-	CCharacter *pTargetChr = GameServer()->m_World.IntersectCharacter(PrevPos, ColPos, m_Freeze ? 1.0f : 6.0f, ColPos, pOwnerChar, m_Owner);
+	CCharacter *pTargetChr = nullptr;
```
Comment (Learath2) [inline src/game/server/entities/projectile.cpp]: 0 in tw code.

**Comment by heinrich5991** [comment]
What's the forward-compatibility story for this?

Do we now guarantee to keep this feature the same forever, or do we mark it as "beta" and will break maps if we find bugs?

**Comment by Learath2** [comment]
I think we should break maps for bugs we find for like a month or two. It's a new feature and it'll be tested by many.

**Comment by Ryozuki** [comment]
The feature is in beta for now, no maps will be accepted from map testing for now, which allows us to not freeze the feature gameplaywise.

**Pull request merged**